### ISSUE TITLE
Text Boundary Detection — Task Guide

### ISSUE THREAD

**Issue by gangagyatso4364** [OpenPecha/BEC#275]
# Project Plan: Tibetan Text Boundary Detection

## 1. Goal

Detect **structural boundaries** in Tibetan book or page content by classifying segments into section types such as **title**, **header**, **body**, **colophon**, and **metadata**. The codebase compares five approaches (A–E) on the same labeled CSV.

---

## 2. Approaches Overview

| | A | B | C | D | E |
|---|---|---|---|---|---|
| **Name** | Semi-rule-based ML | Gemini (few-shot) | Claude (few-shot) | Tibetan RoBERTa (fine-tuned) | mmBERT (fine-tuned) |
| **Method** | RandomForest on statistical + PHD/MLE features | Google Gemini prompting | Anthropic Claude prompting | Sequence classification | Sequence classification |
| **Base model** | `roberta-base` (embeddings only) | Gemini API | Claude API | [`openpecha/tibetan_RoBERTa_S_e6`](https://huggingface.co/openpecha/tibetan_RoBERTa_S_e6) | [`jhu-clsp/mmBERT-base`](https://huggingface.co/jhu-clsp/mmBERT-base) |
| **Params** | N/A (RF) | API | API | ~83M (6-layer RoBERTa) | ~307M (ModernBERT) |
| **Typical train settings** | N/A | N/A | N/A | `--max-length 256` (default in code) | `--max-length 512` (default in `approach_e.finetune`) |
| **Cost** | Free | API usage | API usage | Free after training | Free after training |
| **Env / keys** | None | `GOOGLE_API_KEY` | `ANTHROPIC_API_KEY` | None | None |
| **GPU** | Optional (faster embeddings) | N/A | N/A | Recommended | Recommended (larger model) |
| **Benchmark flag** | `semi-ai` | `gemini` | `claude` | `roberta` + `--model-dir-d` | `mmbert` + `--model-dir-e` |

### When to use which

| Scenario | Suggested approach |
|----------|-------------------|
| No API, quick baseline, modest labeled data | **A** |
| Very few labels, API available | **B** or **C** |
| Tibetan-centric accuracy, enough data + GPU | **D** |
| Multilingual robustness, large GPU | **E** |
| Final comparison | Run **`python -m src.benchmark --approaches all`** with keys and model directories set |

---

## 3. Completion Criteria

Use this as a project checklist (tick as you complete work on your data).

- [x] **A** — Run on target data; record accuracy / F1.
- [ ] **B** — `GOOGLE_API_KEY` set; run and record metrics.
- [ ] **C** — `ANTHROPIC_API_KEY` set; run and record metrics.
- [ ] **D** — Fine-tune on Tibetan labels; save under e.g. `models/tibetan_roberta_finetuned`; evaluate with `--model-dir-d`.
- [ ] **E** — Fine-tune; save under e.g. `models/mmbert_finetuned`; evaluate with `--model-dir-e`.
- [ ] **Decision** — Choose production approach(es) by accuracy, latency, cost, and ops constraints.

---

### 4.0 Data and labels

- [ ] **Label schema** — Keep section types consistent (e.g. title, header, body, colophon, metadata); document guidelines for annotators.
- [ ] **Sources** — Tibetan texts across genres as needed (religious, historical, modern, etc.).
- [ ] **Annotation** — Spreadsheet, Label Studio, or CSV; UTF-8 Tibetan script.
- [ ] **Pilot** — Small labeled set to validate schema and agreement.
- [ ] **Scale** — Grow toward hundreds of examples for stable **D**/**E** fine-tuning.
- [ ] **Quality** — Class balance, spot checks, encoding checks.
- [ ] **Production CSV** — e.g. `data/tibetan_book_data.csv` with at least `text`, `section_type`; optional `source`, `page`.

**Example data in repo:** `python -m src.create_dataset` writes `data/tibetan_example_data.csv` (30 rows, five section types) and `data/tibetan_data_template.csv`.

---

### 4.1 Approach A — Semi-rule ML

- [ ] `pip install -r requirements.txt`
- [ ] `python -m src.benchmark --csv data/tibetan_example_data.csv --approaches semi-ai`
- [ ] Repeat with your full Tibetan CSV when ready.

---

### 4.2 Approach B — Gemini

- [ ] `pip install google-generativeai` (listed in requirements)
- [ ] Set `GOOGLE_API_KEY`
- [ ] `python -m src.benchmark --csv <your.csv> --approaches gemini`

---

### 4.3 Approach C — Claude

- [ ] `pip install anthropic` (listed in requirements)
- [ ] Set `ANTHROPIC_API_KEY`
- [ ] `python -m src.benchmark --csv <your.csv> --approaches claude`

---

### 4.4 Approach D — Tibetan RoBERTa

- [ ] `transformers>=4.48.0` (see `requirements.txt`)
- [ ] Fine-tune:
  ```bash
  python -m src.approach_d.finetune --csv data/tibetan_book_data.csv ^
    --model-name openpecha/tibetan_RoBERTa_S_e6 ^
    --output-dir models/tibetan_roberta_finetuned ^
    --max-length 256 --epochs 5 --batch-size 8 --gpu
  ```
- [ ] Evaluate:
  ```bash
  python -m src.benchmark --csv data/tibetan_book_data.csv --approaches roberta ^
    --model-dir-d models/tibetan_roberta_finetuned
  ```

---

### 4.5 Approach E — mmBERT

- [ ] Optional: `flash-attn` on GPU (see `requirements.txt` comments)
- [ ] Fine-tune (defaults tuned for mmBERT):
  ```bash
  python -m src.approach_e.finetune --csv data/tibetan_book_data.csv ^
    --output-dir models/mmbert_finetuned ^
    --epochs 5 --batch-size 4 --gpu
  ```
  (`--model-name` and `--max-length 512` are defaults in this entry point.)
- [ ] Evaluate:
  ```bash
  python -m src.benchmark --csv data/tibetan_book_data.csv --approaches mmbert ^
    --model-dir-e models/mmbert_finetuned
  ```

---

### 4.6 Full benchmark (A–E)

```bash
# PowerShell: set keys for the session, then:
python -m src.benchmark --csv data/tibetan_book_data.csv --approaches all `
  --model-dir-d models/tibetan_roberta_finetuned `
  --model-dir-e models/mmbert_finetuned `
  --output-csv results_all.csv
```

Approaches **D** and **E** are skipped if the corresponding `--model-dir-*` is missing (a warning is printed).

---

### 4.7 Reporting

- [ ] Store runs in CSV (`--output-csv`) and note date, data version, and hyperparameters.
- [ ] Short write-up: recommended approach(s) and tradeoffs.
- [ ] Keep **README.md** as the user-facing entry (it matches the current CLI).

---

## 5. Collaborators (optional)

- @kaldan007 — annotation / evaluation (if applicable)
- @ta4tsering — annotation / evaluation (if applicable)

---

## 6. Repository layout (current)

```
text-boundary-detection/
├── README.md
├── project_plan.md
├── requirements.txt
├── Book_Boundary_Benchmark.ipynb
├── src/
│   ├── benchmark.py              # CLI: python -m src.benchmark
│   ├── create_dataset.py         # Tibetan example CSV + template
│   ├── common/                   # metrics, evaluate_approach, comparison tables
│   ├── approach_a/               # Semi-rule ML + features
│   ├── approach_b/               # Gemini
│   ├── approach_c/               # Claude
│   ├── approach_d/               # Tibetan RoBERTa detector + finetune.py
│   └── approach_e/               # mmBERT detector + finetune.py
├── data/
│   ├── tibetan_example_data.csv  # after running create_dataset
│   └── tibetan_data_template.csv
└── models/                       # created when you fine-tune (gitignored if needed)
    ├── tibetan_roberta_finetuned/
    └── mmbert_finetuned/
```

---

## 7. CLI reference (summary)

| Command | Purpose |
|---------|---------|
| `python -m src.create_dataset` | Generate example Tibetan CSV + template |
| `python -m src.benchmark --help` | All benchmark flags |
| `python -m src.approach_d.finetune --help` | Fine-tune D (any HF model via `--model-name`) |
| `python -m src.approach_e.finetune --help` | Fine-tune E (mmBERT defaults) |

Use **`--model-dir-d`** and **`--model-dir-e`** for benchmark evaluation of fine-tuned checkpoints (not a single `--model-dir`).

**Comment by lobsam** [comment]
@gangagyatso4364  Multilingual roberta base finetune

**Comment by lobsam** [comment]
@gangagyatso4364  is working on approach A, D and E

**Comment by ta4tsering** [comment]
didnt follow through with this plan

**Comment by webuddhist-bot1** [comment]
## 📋 Card Review

**Card Type:** Developer  
**Status:** Ready

### Good Points
- Description is clear and actionable
- Scope/Subtasks defined
- Completion Criteria present
